base/acme/src/org/dogtagpki/acme/database/LDAPDatabase.java

```diff
@@ -0,0 +1,607 @@
+        LDAPEntry entry = ldapGet(dn);
+        if (entry == null) return null;
+
+        ACMEAccount account = new ACMEAccount();
+        account.setID(accountID);
+
+        LDAPAttribute attrStatus = entry.getAttribute(ATTR_STATUS);
+        account.setStatus(attrStatus.getStringValues().nextElement());
+
+        LDAPAttribute attrKey = entry.getAttribute(ATTR_ACCOUNT_KEY);
+        account.setJWK(JWK.fromJSON(attrKey.getStringValues().nextElement()));
+
+        // TODO orders ?
+
+        // account was not created unless ToS were agreed
+        account.setTermsOfServiceAgreed(true);
+
+        return account;
+    }
+
+    public void addAccount(ACMEAccount account) throws Exception {
+        LDAPAttribute[] attrs = {
+            new LDAPAttribute(ATTR_OBJECTCLASS, OBJ_ACCOUNT),
+            new LDAPAttribute(ATTR_ACCOUNT_ID, account.getID()),
+            new LDAPAttribute(ATTR_ACCOUNT_KEY, account.getJWK().toJSON()),
+            new LDAPAttribute(ATTR_STATUS, account.getStatus())
+        };
+        LDAPAttributeSet attrSet = new LDAPAttributeSet(attrs);
+        String[] contacts = account.getContact();
+        for (int i = 0; i < contacts.length; i++) {
+            attrSet.add(new LDAPAttribute(ATTR_ACCOUNT_CONTACT, contacts[i]));
+        }
+
+        String dn = ATTR_ACCOUNT_ID + "=" + account.getID() + "," + RDN_ACCOUNT + "," + basedn;
+        LDAPEntry entry = new LDAPEntry(dn, attrSet);
+        ldapAdd(entry);
+    }
+
+    /**
+     * Update account status.  Assume that AccountService has validated all data
+     * and just perform the update.
+     */
+    public void updateAccount(ACMEAccount account) throws Exception {
+        String dn = ATTR_ACCOUNT_ID + "=" + account.getID() + "," + RDN_ACCOUNT + "," + basedn;
+        LDAPModificationSet mods = new LDAPModificationSet();
+        mods.add(
+            LDAPModification.REPLACE,
+            new LDAPAttribute(ATTR_STATUS, account.getStatus())
+        );
+        mods.add(
+            LDAPModification.REPLACE,
+            new LDAPAttribute(ATTR_ACCOUNT_CONTACT, account.getContact())
+        );
+        ldapModify(dn, mods);
+    }
+
+    public ACMEOrder getOrder(String orderID) throws Exception {
+        String dn = ATTR_ORDER_ID + "=" + orderID + "," + RDN_ORDER + "," + basedn;
+        LDAPEntry entry = ldapGet(dn);
+        if (entry == null) return null;
+        return loadOrder(entry);
+    }
+
+    public ACMEOrder getOrderByAuthorization(String authzID) throws Exception {
+        Vector<LDAPEntry> entries = ldapSearch(
+            RDN_ORDER + "," + basedn,
+            "(&(" + ATTR_OBJECTCLASS + "=" + OBJ_ORDER +
+                ")(" + ATTR_AUTHORIZATION_ID + "=" + authzID + "))"
+        );
+        if (entries.size() < 1) return null;
+
+        // In the ACME data model, an authorization may be associated with
+        // zero, one or multiple orders.  But here we ignore this and just
+        // return the first matching entry.
```

Comment:
Not quite sure I follow - do you mean this method only return a pending authz?

If so I see two options:

1. Change the function to:
```java
public Collection<ACMEOrder> getOrdersByAuthorization(String authzID)
```
and return *all* orders, allowing the caller to filter it down to pending order(s).

or,

2. Change it to `getPendingOrder(s)ByAuthorization` and fliter it down to pending authorization(s) only, within this method.

What do you think?

Reply:
Yeah, option #2 might be better since `ACMEChallengeService` can use the results immediately without any additional filtering. Regardless, `ACMEChallengeService` will have to process multiple orders instead of just one.

Do you want me to make this change or do you prefer doing it? It will affect the other database plugins.

Reply:
Hm, I'll have a go at it.  If I get stuck on the PostgreSQL impl I'll kick it over to you then :)